File: NicksCookbook/ClaudeService.swift
import Foundation
import UIKit

// MARK: - Claude Service

actor ClaudeService {
    private let baseURL = "https://api.anthropic.com/v1/messages"
    private let apiKey: String

    // Raw conversation history stored as dictionaries for flexible message building
    private var conversationHistory: [[String: Any]] = []

    init(apiKey: String) {
        self.apiKey = apiKey
    }

    // MARK: - Public Interface

    /// Analyzes a fridge image and returns identified ingredients + 5 recipes.
    func analyzeImageAndGetRecipes(imageData: Data) async throws -> (ingredients: [String], recipes: [Recipe]) {
        conversationHistory = []

        let base64Image = imageData.base64EncodedString()

        let imageContent: [String: Any] = [
            "type": "image",
            "source": [
                "type": "base64",
                "media_type": "image/jpeg",
                "data": base64Image
            ] as [String: Any]
        ]

        let textContent: [String: Any] = [
            "type": "text",
            "text": initialRecipePrompt()
        ]

        let userMessage: [String: Any] = [
            "role": "user",
            "content": [imageContent, textContent]
        ]

        conversationHistory.append(userMessage)

        let responseText = try await callClaude(systemPrompt: systemPrompt())
        conversationHistory.append(["role": "assistant", "content": responseText])

        return try parseRecipeResponse(responseText)
    }

    /// Analyzes a plain-text ingredient list and returns 5 recipes (no image required).
    func analyzeTextIngredients(ingredients: [String]) async throws -> (ingredients: [String], recipes: [Recipe]) {
        conversationHistory = []

        let ingredientList = ingredients.joined(separator: ", ")
        let textContent: [String: Any] = [
            "type": "text",
            "text": textIngredientPrompt(ingredientList)
        ]

        let userMessage: [String: Any] = [
            "role": "user",
            "content": [textContent]
        ]

        conversationHistory.append(userMessage)

        let responseText = try await callClaude(systemPrompt: systemPrompt())
        conversationHistory.append(["role": "assistant", "content": responseText])

        return try parseRecipeResponse(responseText)
    }

    /// Requests 5 more recipes, optionally considering user feedback.
    /// `currentIngredients` ensures any edits the user made to the ingredient list are honoured.
    func getMoreRecipes(feedback: String?, currentIngredients: [String]) async throws -> [Recipe] {
        let feedbackClause: String
        if let feedback = feedback, !feedback.isEmpty {
            feedbackClause = " taking this feedback into account: \"\(feedback)\""
        } else {
            feedbackClause = ""
        }

        let ingredientList = currentIngredients.isEmpty
            ? "the same ingredients identified earlier"
            : currentIngredients.joined(separator: ", ")

        let prompt = """
        Please suggest 5 more different recipes\(feedbackClause). Use these ingredients: \(ingredientList). Do NOT repeat any recipes already provided. Aim for variety — different cuisines, cooking methods, and meal types.

        Respond ONLY with valid JSON in this exact format:
        {
            "identified_ingredients": [],
            "recipes": [
                {
                    "name": "Recipe Name",
                    "description": "Brief, appetizing description",
                    "ingredients": ["amount ingredient", ...],
                    "instructions": ["Step 1: ...", "Step 2: ...", ...],
                    "prep_time": "X minutes",
                    "cook_time": "X minutes",
                    "servings": 4
                }
            ]
        }
        """

        let userMessage: [String: Any] = [
            "role": "user",
            "content": prompt
        ]
        conversationHistory.append(userMessage)

        let responseText = try await callClaude(systemPrompt: nil)
        conversationHistory.append(["role": "assistant", "content": responseText])

        let (_, recipes) = try parseRecipeResponse(responseText)
        return recipes
    }

    // MARK: - Private Helpers

    private func systemPrompt() -> String {
        """
        You are Nick's Cookbook — a friendly, expert culinary assistant. Your specialty is identifying ingredients in fridge photos and crafting creative, delicious recipes from whatever is available.

        Guidelines:
        - Be practical: use visible ingredients as the foundation, supplemented by common pantry staples (oil, salt, pepper, basic spices, pasta, rice, etc.)
        - Be creative: suggest varied cuisines and cooking styles
        - Be encouraging: write appetizing descriptions that make recipes sound irresistible
        - Always respond with valid JSON only — no extra text before or after the JSON object
        """
    }

    private func textIngredientPrompt(_ ingredientList: String) -> String {
        """
        I have these ingredients available: \(ingredientList)

        Please suggest exactly 5 delicious, practical recipes using these ingredients (supplemented by common pantry staples like oil, salt, pepper, pasta, rice, etc.).

        Respond ONLY with a valid JSON object in this exact format — no other text:
        {
            "identified_ingredients": [\(ingredientList.split(separator: ",").map { "\"\($0.trimmingCharacters(in: .whitespaces))\"" }.joined(separator: ", "))],
            "recipes": [
                {
                    "name": "Recipe Name",
                    "description": "Brief, appetizing description (2-3 sentences)",
                    "ingredients": ["2 cups ingredient1", "1 tbsp ingredient2", ...],
                    "instructions": ["Step 1: ...", "Step 2: ...", ...],
                    "prep_time": "15 minutes",
                    "cook_time": "25 minutes",
                    "servings": 4
                }
            ]
        }

        Make sure:
        - Each recipe uses primarily the listed ingredients
        - Instructions are clear and numbered
        - Times are realistic
        - Variety across the 5 recipes (different meals, cuisines, complexity levels)
        """
    }

    private func initialRecipePrompt() -> String {
        """
        Please analyze this fridge photo carefully and:
        1. Identify all visible ingredients (be specific — note quantities where visible)
        2. Suggest exactly 5 delicious, practical recipes using these ingredients

        Respond ONLY with a valid JSON object in this exact format — no other text:
        {
            "identified_ingredients": ["ingredient1", "ingredient2", ...],
            "recipes": [
                {
                    "name": "Recipe Name",
                    "description": "Brief, appetizing description (2-3 sentences)",
                    "ingredients": ["2 cups ingredient1", "1 tbsp ingredient2", ...],
                    "instructions": ["Step 1: ...", "Step 2: ...", ...],
                    "prep_time": "15 minutes",
                    "cook_time": "25 minutes",
                    "servings": 4
                }
            ]
        }

        Make sure:
        - Each recipe uses primarily the visible fridge ingredients
        - Instructions are clear and numbered
        - Times are realistic
        - Variety across the 5 recipes (different meals, cuisines, complexity levels)
        """
    }

    private func callClaude(systemPrompt: String?) async throws -> String {
        guard !apiKey.isEmpty else {
            throw ClaudeError.missingAPIKey
        }

        var requestBody: [String: Any] = [
            "model": "claude-opus-4-6",
            "max_tokens": 4096,
            "thinking": ["type": "adaptive"] as [String: Any],
            "messages": conversationHistory
        ]

        if let system = systemPrompt {
            requestBody["system"] = system
        }

        let jsonData = try JSONSerialization.data(withJSONObject: requestBody)

        var request = URLRequest(url: URL(string: baseURL)!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.httpBody = jsonData
        request.timeoutInterval = 180

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw ClaudeError.invalidResponse
        }

        if httpResponse.statusCode != 200 {
            let errorBody = String(data: data, encoding: .utf8) ?? "No error body"
            throw ClaudeError.apiError(statusCode: httpResponse.statusCode, message: errorBody)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let content = json["content"] as? [[String: Any]] else {
            throw ClaudeError.invalidResponse
        }

        // Filter for text blocks (skip thinking blocks)
        guard let textBlock = content.first(where: { $0["type"] as? String == "text" }),
              let text = textBlock["text"] as? String else {
            throw ClaudeError.invalidResponse
        }

        return text
    }

    private func parseRecipeResponse(_ response: String) throws -> (ingredients: [String], recipes: [Recipe]) {
        // Extract the JSON object from the response text
        let jsonString = extractJSON(from: response)

        guard let jsonData = jsonString.data(using: .utf8) else {
            throw ClaudeError.parseError("Could not convert response to data")
        }

        do {
            let analysisResponse = try JSONDecoder().decode(FridgeAnalysisResponse.self, from: jsonData)
            return (analysisResponse.identifiedIngredients, analysisResponse.recipes)
        } catch {
            throw ClaudeError.parseError("JSON decode failed: \(error.localizedDescription)\n\nResponse was: \(jsonString.prefix(500))")
        }
    }

    /// Extracts the outermost JSON object from a string that may contain extra text.
    private func extractJSON(from text: String) -> String {
        var depth = 0
        var startIndex: String.Index?
        var endIndex: String.Index?

        for (idx, char) in text.enumerated() {
            let index = text.index(text.startIndex, offsetBy: idx)
            if char == "{" {
                if depth == 0 {
                    startIndex = index
                }
                depth += 1
            } else if char == "}" {
                depth -= 1
                if depth == 0, let start = startIndex {
                    endIndex = text.index(after: index)
                    return String(text[start..<endIndex!])
                }
            }
        }

        return text
    }
}

// MARK: - Errors

enum ClaudeError: LocalizedError {
    case missingAPIKey
    case invalidResponse
    case apiError(statusCode: Int, message: String)
    case parseError(String)

    var errorDescription: String? {
        switch self {
        case .missingAPIKey:
            return "No API key configured. Please add your Anthropic API key in Settings."
        case .invalidResponse:
            return "Received an invalid response from the Claude API."
        case .apiError(let code, let message):
            return "API Error \(code): \(message)"
        case .parseError(let detail):
            return "Failed to parse recipe data: \(detail)"
        }
    }
}

File: web/src/services/claudeService.ts
import type { Recipe } from '../types'

const API_URL = 'https://api.anthropic.com/v1/messages'

type MsgRole = 'user' | 'assistant'
type RawMessage = { role: MsgRole; content: unknown }

// ---------------------------------------------------------------------------
// Image compression utility
// ---------------------------------------------------------------------------

export async function imageFileToBase64(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image()
    const url = URL.createObjectURL(file)

    img.onload = () => {
      const MAX = 1280
      let w = img.naturalWidth
      let h = img.naturalHeight
      if (w > MAX || h > MAX) {
        if (w >= h) { h = Math.round(h * MAX / w); w = MAX }
        else        { w = Math.round(w * MAX / h); h = MAX }
      }
      const canvas = document.createElement('canvas')
      canvas.width = w
      canvas.height = h
      const ctx = canvas.getContext('2d')!
      ctx.drawImage(img, 0, 0, w, h)
      URL.revokeObjectURL(url)
      // strip the "data:image/jpeg;base64," prefix
      resolve(canvas.toDataURL('image/jpeg', 0.8).split(',')[1])
    }

    img.onerror = () => {
      URL.revokeObjectURL(url)
      reject(new Error('Could not load image'))
    }
    img.src = url
  })
}

// ---------------------------------------------------------------------------
// ClaudeService
// ---------------------------------------------------------------------------

export class ClaudeService {
  private readonly apiKey: string
  private history: RawMessage[] = []

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  async analyzeImage(base64: string): Promise<{ ingredients: string[]; recipes: Recipe[] }> {
    this.history = []

    const msg: RawMessage = {
      role: 'user',
      content: [
        {
          type: 'image',
          source: { type: 'base64', media_type: 'image/jpeg', data: base64 },
        },
        { type: 'text', text: INITIAL_PROMPT },
      ],
    }
    this.history.push(msg)

    const response = await this.call(SYSTEM_PROMPT)
    this.history.push({ role: 'assistant', content: response })
    return this.parse(response)
  }

  async analyzeTextIngredients(ingredients: string[]): Promise<{ ingredients: string[]; recipes: Recipe[] }> {
    this.history = []

    const list = ingredients.join(', ')
    const msg: RawMessage = {
      role: 'user',
      content: textIngredientPrompt(list),
    }
    this.history.push(msg)

    const response = await this.call(SYSTEM_PROMPT)
    this.history.push({ role: 'assistant', content: response })
    return this.parse(response)
  }

  async getMoreRecipes(feedback: string | null, currentIngredients: string[]): Promise<Recipe[]> {
    const feedbackClause = feedback?.trim()
      ? ` taking this feedback into account: "${feedback.trim()}"`
      : ''
    const list = currentIngredients.length
      ? currentIngredients.join(', ')
      : 'the same ingredients identified earlier'

    const prompt = `Please suggest 5 more different recipes${feedbackClause}. Use these ingredients: ${list}. Do NOT repeat any recipes already provided. Aim for variety — different cuisines, cooking methods, and meal types.

Respond ONLY with valid JSON in this exact format:
{
  "identified_ingredients": [],
  "recipes": [
    {
      "name": "Recipe Name",
      "description": "Brief appetizing description",
      "ingredients": ["amount ingredient", ...],
      "instructions": ["Step 1: ...", ...],
      "prep_time": "X minutes",
      "cook_time": "X minutes",
      "servings": 4
    }
  ]
}`

    this.history.push({ role: 'user', content: prompt })
    const response = await this.call(null)
    this.history.push({ role: 'assistant', content: response })
    const { recipes } = this.parse(response)
    return recipes
  }

  // ── Private ───────────────────────────────────────────────────────────────

  private async call(systemPrompt: string | null): Promise<string> {
    const body: Record<string, unknown> = {
      model: 'claude-opus-4-6',
      max_tokens: 4096,
      thinking: { type: 'adaptive' },
      messages: this.history,
    }
    if (systemPrompt) body.system = systemPrompt

    const res = await fetch(API_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': this.apiKey,
        'anthropic-version': '2023-06-01',
        // Required for direct browser access
        'anthropic-dangerous-direct-browser-access': 'true',
      },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(180_000),
    })

    if (!res.ok) {
      const body = await res.text()
      throw new Error(`API Error ${res.status}: ${body}`)
    }

    const data = await res.json() as { content: Array<{ type: string; text?: string }> }
    const textBlock = data.content.find(b => b.type === 'text')
    if (!textBlock?.text) throw new Error('No text response from Claude')
    return textBlock.text
  }

  private parse(response: string): { ingredients: string[]; recipes: Recipe[] } {
    const jsonStr = extractJSON(response)
    const data = JSON.parse(jsonStr) as {
      identified_ingredients?: string[]
      recipes: Array<{
        name: string
        description: string
        ingredients: string[]
        instructions: string[]
        prep_time: string
        cook_time: string
        servings: number
      }>
    }

    const recipes: Recipe[] = data.recipes.map(r => ({
      id: crypto.randomUUID(),
      name: r.name,
      description: r.description,
      ingredients: r.ingredients,
      instructions: r.instructions,
      prepTime: r.prep_time,
      cookTime: r.cook_time,
      servings: r.servings,
    }))

    return { ingredients: data.identified_ingredients ?? [], recipes }
  }
}

// ---------------------------------------------------------------------------
// Helpers / prompts
// ---------------------------------------------------------------------------

function extractJSON(text: string): string {
  let depth = 0, start = -1
  for (let i = 0; i < text.length; i++) {
    if (text[i] === '{') {
      if (depth === 0) start = i
      depth++
    } else if (text[i] === '}') {
      depth--
      if (depth === 0 && start !== -1) return text.slice(start, i + 1)
    }
  }
  return text
}

function textIngredientPrompt(list: string): string {
  const quoted = list.split(',').map(s => `"${s.trim()}"`).join(', ')
  return `I have these ingredients available: ${list}

Please suggest exactly 5 delicious, practical recipes using these ingredients (supplemented by common pantry staples like oil, salt, pepper, pasta, rice, etc.).

Respond ONLY with a valid JSON object in this exact format — no other text:
{
  "identified_ingredients": [${quoted}],
  "recipes": [
    {
      "name": "Recipe Name",
      "description": "Brief, appetizing description (2-3 sentences)",
      "ingredients": ["2 cups ingredient1", "1 tbsp ingredient2", ...],
      "instructions": ["Step 1: ...", "Step 2: ...", ...],
      "prep_time": "15 minutes",
      "cook_time": "25 minutes",
      "servings": 4
    }
  ]
}

Make sure variety across the 5 recipes — different meals, cuisines, complexity levels.`
}

const SYSTEM_PROMPT = `You are Nick's Cookbook — a friendly, expert culinary assistant. Your specialty is identifying ingredients and crafting creative, delicious recipes from whatever is available.

Guidelines:
- Be practical: use visible ingredients as the foundation, supplemented by common pantry staples
- Be creative: suggest varied cuisines and cooking styles
- Be encouraging: write appetizing descriptions that make recipes sound irresistible
- Always respond with valid JSON only — no extra text before or after the JSON object`

const INITIAL_PROMPT = `Please analyze this fridge photo carefully and:
1. Identify all visible ingredients (be specific — note quantities where visible)
2. Suggest exactly 5 delicious, practical recipes using these ingredients

Respond ONLY with a valid JSON object in this exact format — no other text:
{
  "identified_ingredients": ["ingredient1", "ingredient2", ...],
  "recipes": [
    {
      "name": "Recipe Name",
      "description": "Brief, appetizing description (2-3 sentences)",
      "ingredients": ["2 cups ingredient1", "1 tbsp ingredient2", ...],
      "instructions": ["Step 1: ...", "Step 2: ...", ...],
      "prep_time": "15 minutes",
      "cook_time": "25 minutes",
      "servings": 4
    }
  ]
}

Make sure variety across the 5 recipes — different meals, cuisines, complexity levels.`
